File: chronos-mcp-ts/src/graph.ts
import { v4 as uuidv4 } from 'uuid';

/**
 * Evidence to trace back the origin of a data point, eliminating AI hallucinations.
 */
export interface SourceEvidence {
  /** Unique identifier of the source document/report */
  sourceId: string;
  /** The original text snippet from which the data was extracted */
  textSnippet: string;
  /** The publication or occurrence time of the source */
  timestamp: Date;
}

/**
 * The Universal 4D Tuple: S-P-O-T
 * (Subject, Predicate, Object, Time)
 *
 * This disruptive model breaks the boundaries between traditional property graphs
 * and time-series databases by representing everything as a temporal edge.
 */
export interface FourDTuple {
  /** Unique identifier of the tuple */
  id: string;
  /** The subject entity, e.g., 'CATL', 'Apple' */
  subject: string;
  /** The relation or property, e.g., 'gross margin', 'invests_in' */
  predicate: string;
  /** The object entity or value, e.g., '28.5', 'Solid State Battery' */
  object: string;
  /** The exact time when this fact is valid or occurred */
  time: Date;
  /** The unit if the object is a numeric value, e.g., '%', 'USD' */
  unit?: string;
  /** Source evidence to guarantee traceability */
  evidence?: SourceEvidence;
}

/**
 * Universal 4D Hypergraph Memory Engine.
 *
 * A pure in-memory O(1) indexing engine that natively supports Temporal Knowledge Graphs.
 * It indexes tuples by Subject, Predicate, and Object to allow instant multidimensional queries.
 */
export class UniversalHypergraph {
  public tuples: Record<string, FourDTuple> = {};
  private sIndex: Record<string, string[]> = {};
  private pIndex: Record<string, string[]> = {};
  private oIndex: Record<string, string[]> = {};

  /**
   * Insert a FourDTuple into the hypergraph and update indexes.
   *
   * @param tup - The tuple to insert.
   * @returns The unique ID of the inserted tuple.
   */
  insert(tup: FourDTuple): string {
    if (!tup.id) {
      tup.id = `tup_${uuidv4().substring(0, 8)}`;
    }
    this.tuples[tup.id] = tup;

    if (!this.sIndex[tup.subject]) this.sIndex[tup.subject] = [];
    this.sIndex[tup.subject].push(tup.id);

    if (!this.pIndex[tup.predicate]) this.pIndex[tup.predicate] = [];
    this.pIndex[tup.predicate].push(tup.id);

    if (!this.oIndex[tup.object]) this.oIndex[tup.object] = [];
    this.oIndex[tup.object].push(tup.id);

    return tup.id;
  }

  /**
   * Query the Universal Hypergraph across any combination of dimensions.
   *
   * @param params - The query filters.
   * @returns A chronologically sorted array of matching tuples.
   */
  query(params: {
    subject?: string;
    predicate?: string;
    object?: string;
    startTime?: Date;
    endTime?: Date;
  }): FourDTuple[] {
    let candidates: Set<string> | null = null;

    const intersect = (index: Record<string, string[]>, key: string) => {
      const currentSet = new Set(index[key] || []);
      if (candidates === null) {
        candidates = currentSet;
      } else {
        candidates = new Set([...candidates].filter((x) => currentSet.has(x)));
      }
    };

    if (params.subject) intersect(this.sIndex, params.subject);
    if (params.predicate) intersect(this.pIndex, params.predicate);
    if (params.object) intersect(this.oIndex, params.object);

    if (candidates === null) {
      candidates = new Set(Object.keys(this.tuples));
    }

    const results: FourDTuple[] = [];
    for (const cid of candidates) {
      const t = this.tuples[cid];
      if (params.startTime && t.time < params.startTime) continue;
      if (params.endTime && t.time > params.endTime) continue;
      results.push(t);
    }

    return results.sort((a, b) => a.time.getTime() - b.time.getTime());
  }
}

File: chronos-mcp/src/chronos/graph.py
"""
Core Data Structures for the Chronos-MCP Engine.

This module defines the foundational models for the Universal 4D Hypergraph.
Every piece of information is treated as a 4D Tuple: (Subject, Predicate, Object, Time).
"""

import uuid
import datetime
from typing import Dict, List, Optional, Set
from pydantic import BaseModel, Field


class SourceEvidence(BaseModel):
    """
    Evidence to trace back the origin of a data point, eliminating AI hallucinations.
    """

    source_id: str = Field(
        ..., description="Unique identifier of the source document/report."
    )
    text_snippet: str = Field(
        ..., description="The original text snippet from which the data was extracted."
    )
    timestamp: datetime.datetime = Field(
        ..., description="The publication or occurrence time of the source."
    )


class FourDTuple(BaseModel):
    """
    The Universal 4D Tuple: S-P-O-T
    (Subject, Predicate, Object, Time)

    This disruptive model breaks the boundaries between traditional property graphs
    and time-series databases by representing everything as a temporal edge.
    """

    id: str = Field(default_factory=lambda: f"tup_{uuid.uuid4().hex[:8]}")
    subject: str = Field(..., description="The subject entity, e.g., 'CATL', 'Apple'.")
    predicate: str = Field(
        ..., description="The relation or property, e.g., 'gross margin', 'invests_in'."
    )
    object: str = Field(
        ...,
        description="The object entity or value, e.g., '28.5', 'Solid State Battery'.",
    )
    time: datetime.datetime = Field(
        ..., description="The exact time when this fact is valid or occurred."
    )

    # Metadata
    unit: Optional[str] = Field(
        None, description="The unit if the object is a numeric value, e.g., '%', 'USD'."
    )
    evidence: Optional[SourceEvidence] = Field(
        None, description="Source evidence to guarantee traceability."
    )


class UniversalHypergraph:
    """
    Universal 4D Hypergraph Memory Engine.

    A pure in-memory O(1) indexing engine that natively supports Temporal Knowledge Graphs.
    It indexes tuples by Subject, Predicate, and Object to allow instant multidimensional queries.
    """

    def __init__(self) -> None:
        self.tuples: Dict[str, FourDTuple] = {}

        # O(1) Inverted Indexes
        self._s_index: Dict[str, List[str]] = {}
        self._p_index: Dict[str, List[str]] = {}
        self._o_index: Dict[str, List[str]] = {}

    def insert(self, tup: FourDTuple) -> str:
        """
        Insert a FourDTuple into the hypergraph and update indexes.

        Args:
            tup (FourDTuple): The tuple to insert.

        Returns:
            str: The unique ID of the inserted tuple.
        """
        self.tuples[tup.id] = tup

        if tup.subject not in self._s_index:
            self._s_index[tup.subject] = []
        self._s_index[tup.subject].append(tup.id)

        if tup.predicate not in self._p_index:
            self._p_index[tup.predicate] = []
        self._p_index[tup.predicate].append(tup.id)

        if tup.object not in self._o_index:
            self._o_index[tup.object] = []
        self._o_index[tup.object].append(tup.id)

        return tup.id

    def query(
        self,
        subject: Optional[str] = None,
        predicate: Optional[str] = None,
        object: Optional[str] = None,
        start_time: Optional[datetime.datetime] = None,
        end_time: Optional[datetime.datetime] = None,
    ) -> List[FourDTuple]:
        """
        Query the Universal Hypergraph across any combination of dimensions.

        Args:
            subject (Optional[str]): Filter by subject entity.
            predicate (Optional[str]): Filter by predicate/relation.
            object (Optional[str]): Filter by object entity or value.
            start_time (Optional[datetime.datetime]): Filter tuples occurring on or after this time.
            end_time (Optional[datetime.datetime]): Filter tuples occurring on or before this time.

        Returns:
            List[FourDTuple]: A chronologically sorted list of matching tuples.
        """
        candidates: Optional[Set[str]] = None

        if subject:
            s_set = set(self._s_index.get(subject, []))
            candidates = s_set if candidates is None else candidates.intersection(s_set)
        if predicate:
            p_set = set(self._p_index.get(predicate, []))
            candidates = p_set if candidates is None else candidates.intersection(p_set)
        if object:
            o_set = set(self._o_index.get(object, []))
            candidates = o_set if candidates is None else candidates.intersection(o_set)

        if candidates is None:
            candidates = set(self.tuples.keys())

        results = []
        for cid in candidates:
            t = self.tuples[cid]
            if start_time and t.time < start_time:
                continue
            if end_time and t.time > end_time:
                continue
            results.append(t)

        # Always return chronologically sorted data
        results.sort(key=lambda x: x.time)
        return results
